src/bin/cli.rs

```diff
@@ -0,0 +1,637 @@
+            let specs = args.values_of("spec").unwrap_or_default().collect();
+            ops::uninstall(root, specs, &values(args, "bin"), config)?;
+            Ok(())
+        }
+        ("update", Some(args)) => {
+            let ws = args.workspace(config)?;
+
+            let update_opts = ops::UpdateOptions {
+                aggressive: args.is_present("aggressive"),
+                precise: args.value_of("precise"),
+                to_update: values(args, "package"),
+                config,
+            };
+            ops::update_lockfile(&ws, &update_opts)?;
+            Ok(())
+        }
+        ("verify-project", Some(args)) => {
+            fn fail(reason: &str, value: &str) -> ! {
+                let mut h = HashMap::new();
+                h.insert(reason.to_string(), value.to_string());
+                cargo::print_json(&h);
+                process::exit(1)
+            }
+
+            let mut contents = String::new();
+            let filename = match args.root_manifest(config) {
+                Ok(filename) => filename,
+                Err(e) => fail("invalid", &e.to_string()),
+            };
+
+            let file = File::open(&filename);
+            match file.and_then(|mut f| f.read_to_string(&mut contents)) {
+                Ok(_) => {}
+                Err(e) => fail("invalid", &format!("error reading file: {}", e))
+            };
+            if contents.parse::<toml::Value>().is_err() {
+                fail("invalid", "invalid-format");
+            }
+
+            let mut h = HashMap::new();
+            h.insert("success".to_string(), "true".to_string());
+            cargo::print_json(&h);
+            Ok(())
+        }
+        ("version", _) => {
+            println!("{}", cargo::version());
+            Ok(())
+        }
+        ("yank", Some(args)) => {
+            let registry = args.registry(config)?;
+
+            ops::yank(config,
+                      args.value_of("crate").map(|s| s.to_string()),
+                      args.value_of("vers").map(|s| s.to_string()),
+                      args.value_of("token").map(|s| s.to_string()),
+                      args.value_of("index").map(|s| s.to_string()),
+                      args.is_present("undo"),
+                      registry)?;
+            Ok(())
+        }
+        (cmd, Some(args)) => {
+            if let Some(mut alias) = super::aliased_command(config, cmd)? {
+                alias.extend(args.values_of("").unwrap_or_default().map(|s| s.to_string()));
+                let args = cli()
+                    .setting(AppSettings::NoBinaryName)
+                    .get_matches_from_safe(alias)?;
+                return execute_subcommand(config, args);
+            }
+            let mut ext_args: Vec<&str> = vec![cmd];
+            ext_args.extend(args.values_of("").unwrap_or_default());
+            super::execute_external_subcommand(config, cmd, &ext_args)
+        }
+        _ => Ok(())
```

Comment:
I'm a bit naive to clap (never used it before myself), but what does this case signify? (aka `(cmd, None)`)

Reply:
>never used it before myself

Heh, same here, I think we ought to cc @kbknapp :-) 

>but what does this case signify?

I believe this is for the case "no subcommand is supplied". This case is impossible in theory, because we handle that explicitly in `cli::main`. I think we should just handle it here!

Reply:
@matklad you're correct, it means no subcommand was used. I usually use `("", None) => unreachable!()` if I've set it up so no subcommand is impossible or handled elsewhere.

I'm on mobile but happy to answer any questions yall have ;)